### PR TITLE
Modify ClipConfig API to remove dimension, always clip last dim

### PR DESCRIPTION
**Pull request by mcgibbon**
When we move to tfdatasets, we do not have access to dimension names. In practice we always want to clip the last dimension of an array. This PR modifies clipping configuration to remove the dimension name, and always assume we are clipping the last dimension.

The only training function with more than one non-sample dimension which could possibly receive data with an incorrect dimension ordering is the convolutional model. This model already assumes incoming data has the correct dimensionality, so the clipping assumption of dimension ordering doesn't add any additional danger.

We may consider adding code later to enforce that loader functions return tfdatasets with data of the correct dimensionality.

Refactored public API:
- The `clip` attribute of `PackerConfig` and `ClipConfig` is now `Mapping[Hashable, SliceConfig]`, was `Mapping[Hashable, Mapping[str, SliceConfig]]`.

Significant internal changes:
- refactored clipping functionality to use configuration as `Mapping[Hashable, SliceConfig]`, was `Mapping[Hashable, Mapping[str, SliceConfig]]`.

- [x] Tests added

Resolves #<github issues> [JIRA-TAG]

(Delete this for the commit message)
You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

### CODE REVIEW

File: external/fv3fit/fv3fit/keras/_models/dense.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ class DenseHyperparameters(Hyperparameters):
             each output variable.
         save_model_checkpoints: if True, save one model per epoch when
             dumping, under a 'model_checkpoints' subdirectory
-        clip_config: configuration of dataset packing.
+        clip_config: configuration of input and output clipping
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/dense.py:53]: Can you add to this to mention it is always the last/feature dim?

File: external/fv3fit/fv3fit/keras/_models/shared/clip.py [hunk 2]
```diff
@@ -20,21 +19,12 @@ class ClipConfig(PackerConfig):
             dimension to clip.
     """
 
-    clip: ClipDims = dataclasses.field(default_factory=dict)
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/shared/clip.py:23]: Can you update the docstring above as well? (it says it is 'usually the last dim')

**Review by AnnaKwa — APPROVED** [review]
Thanks, LGTM. Still, I feel a bit wary of having the clipping rely on an assumed dimension ordering that is out of its control since clipping along the wrong dimension could be a silent bug that is hard to catch. I do realize that we already have several other places in the training that rely on dimension ordering to be correct from the loaders though. I don't know where the best place to document the alphabetical dimension ordering behavior would be, but it would be useful to have that communicated somewhere- before our discussion I didn't know this was the case.

Could you add to the PR description to note that the training functions that might make use of this already rely on the assumption that the feature dimension is last?

File: external/fv3fit/fv3fit/keras/_models/dense.py (hunk 1 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/dense.py:53]: Updated.

File: external/fv3fit/fv3fit/keras/_models/shared/clip.py (hunk 2 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/clip.py:23]: Updated.

**Pull request merged**